state/store.go

```diff
@@ -228,129 +229,156 @@ func (store dbStore) Bootstrap(state State) error {
+	paramsInfo, err := store.loadConsensusParamsInfo(retainHeight)
+	if err != nil {
+		return fmt.Errorf("consensus params at height %v not found: %w", retainHeight, err)
+	}
+
+	// As we don't save the consensus params at every height, only when there is a consensus params change,
+	// we must not prune (or save) the last consensus params that the consensus params info at height
+	// is dependent on.
 	if paramsInfo.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
-		keepParams[paramsInfo.LastHeightChanged] = true
+		lastRecordedConsensusParams, err := store.loadConsensusParamsInfo(paramsInfo.LastHeightChanged)
+		if err != nil || lastRecordedConsensusParams.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+			return fmt.Errorf(
+				"couldn't find consensus params at height %d as last changed from height %d: %w",
+				paramsInfo.LastHeightChanged,
+				retainHeight,
+				err,
+			)
+		}
+	}
+
+	// batch delete all the consensus params up to the retain height
+	return store.batchDelete(
+		consensusParamsKey(1),
+		consensusParamsKey(retainHeight),
+		consensusParamsKey(paramsInfo.LastHeightChanged),
+	)
+}
+
+// pruneABCIResponses calls a reverse iterator from base height to retain height batch deleting
+// all abci responses in between
+func (store dbStore) pruneABCIResponses(height int64) error {
+	return store.batchDelete(abciResponsesKey(1), abciResponsesKey(height), nil)
+}
+
+// batchDelete is a generic function for deleting a range of keys in reverse order. It will
+// skip keys that have been
+func (store dbStore) batchDelete(start []byte, end []byte, exception []byte) error {
+	iter, err := store.db.ReverseIterator(start, end)
+	if err != nil {
+		return fmt.Errorf("iterator error: %w", err)
 	}
+	defer iter.Close()
 
 	batch := store.db.NewBatch()
 	defer batch.Close()
-	pruned := uint64(0)
-
-	// We have to delete in reverse order, to avoid deleting previous heights that have validator
-	// sets and consensus params that we may need to retrieve.
-	for h := to - 1; h >= from; h-- {
-		// For heights we keep, we must make sure they have the full validator set or consensus
-		// params, otherwise they will panic if they're retrieved directly (instead of
-		// indirectly via a LastHeightChanged pointer).
-		if keepVals[h] {
-			v, err := loadValidatorsInfo(store.db, h)
-			if err != nil || v.ValidatorSet == nil {
-				vip, err := store.LoadValidators(h)
-				if err != nil {
-					return err
-				}
-
-				pvi, err := vip.ToProto()
-				if err != nil {
-					return err
-				}
-
-				v.ValidatorSet = pvi
-				v.LastHeightChanged = h
-
-				bz, err := v.Marshal()
-				if err != nil {
-					return err
-				}
-				err = batch.Set(validatorsKey(h), bz)
-				if err != nil {
-					return err
-				}
-			}
-		} else {
-			err = batch.Delete(validatorsKey(h))
-			if err != nil {
-				return err
-			}
+
+	pruned := 0
+	for iter.Valid() {
+		key := iter.Key()
+		if exception != nil && bytes.Equal(key, exception) {
```

Comment:
Is key guaranteed to always be non-nil? If so, you can remove the `exception != nil` check.

Reply:
Yeah should be, thanks